### ISSUE TITLE
Incorrect results in Query..

### ISSUE THREAD

**Issue by lloydtabb** [duckdb/duckdb#4766]
### What happens?

The following query should return no results.  It is essentially computing a ratio.  In different versions of DuckDB it returns different results with the same data.

```
WITH __stage0 AS (
  SELECT
    group_set,
    CASE WHEN group_set IN (0,2) THEN
      names."state"
      END as "state__0",
    CASE WHEN group_set=0 THEN
      COALESCE(SUM(names."number"),0)
      END as "population__0",
    CASE WHEN group_set IN (2,1) THEN
      names."name"
      END as "name__2",
    (CASE WHEN group_set=2 THEN
      COALESCE(SUM(names."number"),0)
      END)*1.0/MAX((CASE WHEN group_set=1 THEN
      COALESCE(SUM(names."number"),0)
      END)) OVER (PARTITION BY CASE WHEN group_set IN (2,1) THEN
      names."name"
      END) as "percent_in_state__2"
  FROM usa_names.parquet as names
  CROSS JOIN (SELECT UNNEST(GENERATE_SERIES(0,2,1)) as group_set  ) as group_set
  GROUP BY 1,2,4
  HAVING (group_set<>2 OR (group_set=2 AND (CASE WHEN group_set=2 THEN
    COALESCE(SUM(names."number"),0)
    END)>1000))
)
, __stage1 AS (
  SELECT
    "state__0" as "state",
    MAX(CASE WHEN group_set=0 THEN population__0 END) as "population",
    COALESCE(LIST({
      "name": "name__2", 
      "percent_in_state": "percent_in_state__2"}  ORDER BY  "percent_in_state__2" desc NULLS LAST) FILTER (WHERE group_set=2)[1:40],[]) as "names"
  FROM __stage0
  WHERE group_set NOT IN (1)
  GROUP BY 1
  ORDER BY 2 desc NULLS LAST
)
SELECT 
   base.names[names_0.__row_id]."name" as "name",
   base.names[names_0.__row_id]."percent_in_state" as "percent_in_state"
FROM __stage1 as base
LEFT JOIN (select UNNEST(generate_series(1,
        100000, --
        -- (SELECT genres_length FROM movies limit 1),
        1)) as __row_id) as names_0 ON  names_0.__row_id <= array_length(base."names")
WHERE base.names[names_0.__row_id]."percent_in_state">1
```

It should have no results

![image](https://user-images.githubusercontent.com/1093458/191266919-0c292318-959f-452b-a2e0-9a07a7f2c5ad.png)

Data can be found here:

https://github.com/lloydtabb/name_fiddle

Seems to work fine in WASM version...

https://lloydtabb.github.io/name_fiddle/index.html?q=1+-+Preview+the+table&m=SSA+Names+Database&t=%2F%2F+--%0A%2F%2F+Name%3A+Most+Iconic+names+per+State.%0Aquery%3A+names-%3E+%7B%0A++group_by%3A+state%0A++aggregate%3A+population%0A++nest%3A+names+is+%7B%0A++++group_by%3A+name%0A++++aggregate%3A+%0A++++++percent_in_state+is+population+%2F+exclude%28population%2C+state%29%0A++++limit%3A+40%0A++++having%3A+population+%3E+1000%0A++%7D+%0A%7D%0A-%3E+%7B%0A++++project%3A+names.name%2C+names.percent_in_state%0A++++where%3A+names.percent_in_state+%3E+1%0A%7D%0A

### To Reproduce

instructions above

### OS:

Linux

### DuckDB Version:

5.0

### DuckDB Client:

Malloy

### Full Name:

lloyd tabb

### Affiliation:

Google

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by Mytherin** [comment]
@taniabogatsch could you have a look?

**Comment by taniabogatsch** [comment]
I started looking into this. If an empty result is the desired output, then I did pinpoint the commit that introduced the bug.
2c623978d700443665d519282ee52891e4573a3d is still returning an empty result, whereas d53dc9b62c62275112bcdc3259721c63d5acf1da is the first commit with the bug. 
```
d53dc9b62c62275112bcdc3259721c63d5acf1da is the first bad commit
commit d53dc9b62c62275112bcdc3259721c63d5acf1da
Author: Richard Wesley <hawkfish@electricfish.com>
Date:   Fri Jul 15 13:34:15 2022 -0700

    Issue #4135: Radix Window Masking
    
    Replace the column-by-column partition and peer mask construction
    with a single scan over the radix data using the radix scanning
    iterators from IEJoin.

 src/common/sort/sorted_block.cpp                   |  24 +++
 .../operator/aggregate/physical_window.cpp         | 213 +++++++--------------
 src/execution/operator/join/physical_iejoin.cpp    | 101 +---------
 src/include/duckdb/common/sort/sorted_block.hpp    |  80 ++++++++
 4 files changed, 175 insertions(+), 243 deletions(-)
```
It actually returns this result:
```
┌─────────────┬──────────────────┐
│    name     │ percent_in_state │
├─────────────┼──────────────────┤
│ Christopher │ 17293.0          │
│ Christopher │ 18482.2          │
│ Christopher │ 12614.4          │
│ Christopher │ 11152.8          │
│ Christopher │ 6308.0           │
│ Christopher │ 7237.6           │
│ Christopher │ 4993.0           │
│ Christopher │ 3611.6           │
│ Christopher │ 4808.4           │
│ Christopher │ 5582.6           │
│ Christopher │ 2611.0           │
│ Christopher │ 1220.2           │
│ Christopher │ 897.8            │
│ Christopher │ 1808.8           │
│ Christopher │ 1258.0           │
│ Christopher │ 957.8            │
│ Christopher │ 906.0            │
└─────────────┴──────────────────┘
```
The current master returns this, so there was another change somewhere that affected this query. I can try to find that one too, if we have difficulties finding the bug.
```
┌─────────────┬───────────────────┐
│    name     │ percent_in_state  │
├─────────────┼───────────────────┤
│ Maximillian │ 4.115241635687732 │
└─────────────┴───────────────────┘
```

**Comment by taniabogatsch** [comment]
d53dc9b62c62275112bcdc3259721c63d5acf1da introduced a new method `WindowGlobalHashGroup::ComputeMasks` to compute the validity masks `partition_mask` and `order_mask`.

I created a branch on my fork to isolate the bug better.
https://github.com/taniabogatsch/duckdb/commit/61ea0042c733113c29d92f3a0874c86fd1eaf15f

It seems like `partition_mask` is not created correctly. When swapped with the previous `partition_mask2`, then the query returns the correct result. I tried to find the mismatching index in the validity data, but this for-loop did not throw an exception. I am probably missing something there...
```c++
for (idx_t i = 0; i < entry_count; i++) {
	if (partition_mask.RowIsValid(i) != partition_mask2.RowIsValid(i)) {
		throw InternalException("Inconsistency in validity masks for partition");
	}
}
```
My wild guess is that we need to check for some extra conditions or something when setting the `partition_mask` here in the `ComputeMasks` method.
```c++
if (part_cmp) {
	partition_mask.SetValidUnsafe(curr.GetIndex());
	order_mask.SetValidUnsafe(curr.GetIndex());
} else if (prev.Compare(curr)) {
	order_mask.SetValidUnsafe(curr.GetIndex());
}
```
Sadly, I was not yet able to reduce the query to a smaller reproducible example. Will try that tomorrow and also continue to look into this. Just keeping track of the progress here.

**Comment by hannes** [comment]
Maybe @hawkfish wants to have a look as well?

**Comment by hawkfish** [comment]
The bitmaps differ at seven locations. The first three are at positions 14200, 14201 and 14215; the last four are scattered later on at 15361, 41241, 48616 and 50177.

**Comment by lloydtabb** [comment]
> WHERE base.names[names_0.__row_id]."percent_in_state">1

If you remove the above where clause, you should see results where `percent_in_state` is a value between 0 and 1.  There should be no way for it to exceed 1...

**Comment by taniabogatsch** [comment]
The current master seems to fix the issue, specifically this PR: https://github.com/duckdb/duckdb/commit/94bb7737162c53eb36f9f3f348ebe834d9ba86c7 The percentages in the window are now all below or equal to 1.

What the PR does is that radix sort switches to pdqsort if the section of the prefix that it is sorting contains strings. So it probably didn't fix the issue in the code, but just doesn't use the code for the query any longer. Please correct me if I am wrong @lnkuiper.

These are Richard's findings earlier on the issue:
> [...] it looks to me like the old code was actually wrong. The old sort has some out of order values at a point (14189) where the strings bounce between 11 and 12 characters. “Christopher[e]”. So this may not be a bug.

This seems to me like there were actually two bugs, one was already fixed, but it unearthed this second one (because the percentages where still above 1).

**Comment by taniabogatsch** [comment]
This is the last incorrect commit 7daa9b26cfc2ffa1d1f79547f2fbac16f85883d4.

Here, the `percent_in_state__2` in the query still returns this.
```sql
D SELECT * from __stage0 where percent_in_state__2 > 1;
┌───────────┬──────────┬───────────────┬─────────────┬─────────────────────┐
│ group_set │ state__0 │ population__0 │   name__2   │ percent_in_state__2 │
├───────────┼──────────┼───────────────┼─────────────┼─────────────────────┤
│ 2         │ CA       │               │ Maximillian │ 4.115241635687732   │
└───────────┴──────────┴───────────────┴─────────────┴─────────────────────┘
```

**Comment by lnkuiper** [comment]
I suspect that this is an issue in the sort code. When we have strings that are larger than a certain size, we encode a prefix, sort by the prefix, then break any remaining ties by checking the full string. In a recent PR we've switched to using pdqsort when there strings, so this issue has been now fixed, because the code path containing the error is no longer used for this query.

However, the bug probably still remains, and will be triggered for specific ORDER BY clauses that are a mix of long strings and fixed-size columns. I will look into creating a small reproducible example.

**Comment by lnkuiper** [comment]
I was able to track down the bug. A PR to fix it is open @lloydtabb

**Comment by hannes** [comment]
Good work team, it may have been the first bug that half the team looks at :)

**Comment by Mytherin** [comment]
@lnkuiper this should be fixed by #4816, right?

**Comment by lnkuiper** [comment]
@Mytherin yes, should be fixed!
